src/sentry/migrations/0378_backfill_perf_issue_events_issue_platform.py

```diff
@@ -0,0 +1,648 @@
+        Path(path).mkdir(parents=True, exist_ok=True)
+
+        return f"{path}{path_suffix}.csv"
+
+
+if typing.TYPE_CHECKING:
+    FileBackedProjectProgressType = UserDict[int, ProjectGroupTotalProgress]
+else:
+    FileBackedProjectProgressType = UserDict
+
+
+class FileBackedProjectProgress(FileBackedProjectProgressType):
+    """
+    Holds the file handles for maintaining the progress of migrations for a project. Hopefully, we shouldn't hit
+    the OS limit on open handles since the project with the most performance issues is around 9k. So this class
+    should only be holding at most 2 * 9k file handles before this migration moves on to the next project.
+
+    """
+
+    def __init__(self, project_id: int) -> None:
+        super().__init__()
+        self.project_id = project_id
+
+    def __enter__(self) -> "FileBackedProjectProgress":
+        return self
+
+    def __exit__(self, *args: Any, **kwargs: Any) -> None:
+        try:
+            for progress in self.values():
+                progress.close()
+        except ValueError:
+            pass
+
+    def get_or_init(self, group_id: int) -> ProjectGroupTotalProgress:
+        if self.get(group_id):
+            return self.__getitem__(group_id)
+        else:
+            progress = ProjectGroupTotalProgress(self.project_id, group_id)
+            self.__setitem__(group_id, progress)
+            return progress
+
+
+def backfill_by_project(project_id: int, Group: Any, GroupHash: Any, dry_run: bool) -> None:
+    next_offset: Optional[int] = 0
+
+    # TODO: need to revisit this since we'll be inserting data to search_issues dataset as we backfill
+    #        gonna need to figure out how to ensure we don't process dupes
+    already_processed_events: Mapping[int, typing.Set[str]] = defaultdict(
+        set
+    )  # group_id: Set[event_id]
+
+    # it's probably ok to query all the events here instead of paginating since we should be running this script
+    # closely after we cut-over
+    # for issue_platform_events in _query_issue_platform_events(
+    #     project_id, start=ISSUE_PLATFORM_INGEST_START_DATETIME, end=END_DATETIME
+    # ):
+    #     already_processed_events[issue_platform_events["group_id"]].add(
+    #         issue_platform_events["event_id"]
+    #     )
+
+    with FileBackedProjectProgress(project_id) as previous_progress:
+        while next_offset is not None:
+            rows, next_offset = _query_performance_issue_events(
+                project_id=project_id,
+                start=START_DATETIME,
+                end=END_DATETIME,
+                offset=next_offset,
+            )
+
+            for row in rows:
+                group_id = row["group_id"]
+                event_id = row["event_id"]
+                finish_ts = row["finish_ts"]
+
+                # avoid re-ingesting the same transaction since we'll be dual-writing
+                if event_id not in already_processed_events[group_id]:
```

Comment:
Do we need this check? I'd have thought that we'd only receive each group/event id pair once due to the query + pagination

Reply:
The `already_processed_events` is actually sourced from the `search_issues` dataset. This is to ensure we don't ingest already ingested events that were already persisted in `search_issues`